src/pkt_checker.rs

```diff
@@ -0,0 +1,206 @@
+use std::{
+    collections::{HashMap, LinkedList},
+    sync::{atomic::{AtomicBool, Ordering}, mpsc::Sender, Arc, Mutex, RwLock},
+};
+
+use crate::{
+    op_ctx::ReadOpCtx,
+    recv_pkt_map::RecvPktMap,
+    responser::{RespAckCommand, RespCommand},
+    types::{Msn, Psn},
+    Error,
+};
+
+use log::{error, info};
+
+#[derive(Debug)]
+pub(crate) struct PacketChecker {
+    thread: Option<std::thread::JoinHandle<()>>,
+    stop_flag: Arc<AtomicBool>,
+}
+
+impl PacketChecker {
+    pub(crate) fn new(
+        send_queue: Sender<RespCommand>,
+        recv_pkt_map: Arc<RwLock<HashMap<Msn, Arc<Mutex<RecvPktMap>>>>>,
+        read_op_ctx_map: Arc<RwLock<HashMap<Msn, ReadOpCtx>>>,
+    ) -> Self {
+        let ctx = PacketCheckerContext {
+            send_queue,
+            recv_pkt_map,
+            read_op_ctx_map,
+        };
+        let stop_flag = Arc::new(AtomicBool::new(false));
+        let thread_stop_flag = Arc::clone(&stop_flag);
+        let thread = std::thread::spawn(move || {
+            PacketCheckerContext::working_thread(&ctx, &thread_stop_flag);
+        });
+        Self {
+            thread: Some(thread),
+            stop_flag,
+        }
+    }
+}
+
+impl Drop for PacketChecker {
+    fn drop(&mut self) {
+        self.stop_flag.store(true, Ordering::Relaxed);
+        if let Some(thread) =  self.thread.take(){
+            if let Err(e) = thread.join(){
+                error!("Failed to join the WorkDescPoller thread: {:?}", e);
+            }
+        }
+    }
+}
+
+struct PacketCheckerContext {
+    send_queue: Sender<RespCommand>,
+    recv_pkt_map: Arc<RwLock<HashMap<Msn, Arc<Mutex<RecvPktMap>>>>>,
```

Comment:
Types like this is too complecated, please put them in to structure and abstract some repeated operations into functions. For instance, here you code too much to handle lock things.

Reply:
Yes. There are amount of patterns like `Arc<RwLock<HashMap<K,V>>>`, I will use a type alias. And for struct like `Arc<Mutex<RecvPktMap>>`, I will create an inner struct and wrap it with Arc and Mutex